packages/block-editor/src/components/block-list/block.js

```diff
@@ -401,7 +402,23 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, { select } ) => {
 	return {
 		setAttributes( newAttributes ) {
 			const { clientId } = ownProps;
-			updateBlockAttributes( clientId, newAttributes );
+			const { getBlockRootClientId, getBlockName } = select(
+				'core/block-editor'
+			);
+			let isLocalChange = false;
+			const rootClientId = getBlockRootClientId( clientId );
+			if ( rootClientId ) {
+				// The default save outputs null.
+				// We can have more nuanced heuristics in the future.
+				isLocalChange =
+					select( 'core/blocks' ).getBlockType( getBlockName( rootClientId ) )
```

Comment:
If a block implements its own save that returns null, this logic will fail, right?
I also wonder about back compatibility: Imagining a current block that uses InnerBlocks without any save and listens to save actions (e.g: executes when saved state changes) to save its InnerBlocks using a rest API. Would a block like that break with this change because if its InnerBlocks are changed the post would not become dirty? I think the odds of a block like that existing are low, but at least we would need to create a dev note.

As an alternative solution, what if we use a supports flag to indicate this type of blocks?

Reply:
>If a block implements its own save that returns null, this logic will fail, right?
>I also wonder about back compatibility: Imagining a current block that uses InnerBlocks without any save and listens to save actions (e.g: executes when saved state changes) to save its InnerBlocks using a rest API. Would a block like that break with this change because if its InnerBlocks are changed the post would not become dirty?

Yes, for both, I don't see value in supporting either, though.

>As an alternative solution, what if we use a supports flag to indicate this type of blocks?

It would work, but it would be adding more cruft.

I am not sure. Thoughts @youknowriad?

Reply:
If i understand properly where we're coming from here. i believe all these issues are happening because we mix the blocks of the sub-entity into the blocks of the top entity.

I think we discussed this previously, IMO, we should find a way to stop doing this as it doesn't make sense to mix content from entity A into an edit of entity B. 

One potential idea I shared previously is to have a unique block list in the "editor" store and each entity has its own "blocks" edits in the "core" store.

Reply:
I think the issue is a bit more general than that.

We save serialized content yet detect changes on block objects. Any edit to a block will be dirtying even if its serialization doesn't change.

>One potential idea I shared previously is to have a unique block list in the "editor" store and each entity has its own "blocks" edits in the "core" store.

So like reusable blocks, but resolved into a single blocklist? What do you see as the benefits of that approach over this one?

Reply:
> So like reusable blocks, but resolved into a single blocklist? What do you see as the benefits of that approach over this one?

The benefits is that an entity is dirty only if its block list is dirty (A template block list won't include the blocks from the template area)

Reply:
The same is true in this PR.

So for your suggestion, we would need a new property in block objects that potentially contain a selector that gets a block list for their inner blocks. The editor store would then dynamically resolve a complete block list. However, when the block editor makes an edit, it will mutate the full block list, how would the editor know to not dirty the top-level entity. It seems like we are trading some complexity for another.

Reply:
Agreed no silver bullet here. I think "onChange" of the "controlled" inner blocks might be the place where we dispatch the change to two block lists. Conceptually speaking, I feel like having in an entity of core-data, things that are saved in another entity will just bite us with this kind of hidden bugs.

Reply:
>I think "onChange" of the "controlled" inner blocks might be the place where we dispatch the change to two block lists.

We'll still have the same issues when we nest one of these blocks inside another.

>Conceptually speaking, I feel like having in an entity of core-data, things that are saved in another entity will just bite us with this kind of hidden bugs.

I don't see a conceptual issue with it. They are object references, it's not like the data is duplicated. That's why this PR was so simple to do.